[최용석] Step3 리뷰요청드립니다.

badaelephant opened the pull request:
안녕하세요 원철님.

이번주에는 회사에서 ㅠㅠ 일이 많아서 미션 수행이 많이 늦었네요. 항상 빠르게 리뷰 해주셨는데, 제가 오히려 오래걸려서 죄송하네요. 

이번 Step3에서는 일급컬렉션의 활용과 도메인 로직의 활용을 많이 생각해볼 수 있었습니다. 

작성하면서 조금 궁금했던 것이 한가지 있는데, 바로 DeleteHistories 등의 getter함수에 대한 부분이었습니다. 

Jpa레포지토리를 사용하다보면, findAll, saveAll등의 인자나 결과값이 ArrayList인데요. 그러다보니, 어쩔수없이 DeleteHistories 일급객체를 repository에서 사용할때 getter로 리스트를 가져와서 넣어야 했었습니다. 

첫번째 박재성(자바지기님)님 강의에서 최대한 getter의 사용을 줄이는 것이 좋다고 해서.. 가능하면 getter를 안쓰려고 하는데, 혹시 이렇게 JPA와 관련되서 사용하는 부분은 어쩔수 없는지 혹시 getter를 사용하지 않고 일급컬렉션을 JPA에서 활용하는 좋은 방법이 있다면 알려주시면 감사하겠습니다. 

항상 잘 리뷰해주셔서 감사합니다. ~

Commit:
refactor: Exception 클래스 리팩토링-기존 Exception 클래스 제거 안했음
1. exception 패키지로 이동
2. CannotDeleteException RuntimeException 상속받도록 수정
3. CannotUpdateException 클래스 생성
4. ErrMsg 상수 클래스 생성

Commit:
feat: 일급 컬렉션 객체 생성
1. JPA Embeddable 어노테이션으로 OneToMany 대상 컬렉션을 일급컬렉션으로 클래스 분리
2. 각각의 컬렉션에 update 메서드 추가
3. Answers 클래스에서 답변들 삭제하는 로직을 위임

Commit:
feat: 삭제이력에 of를 통해 Answer, Question 2개 타입으로 삭제이력을 생성하도록함.

Commit:
feat: 답변 클래스 리팩토링
답변 삭제로직 추가
답변 eqals, hashcode로직 오버로딩

Commit:
feat: 질문 클래스 리팩토링
답변 리스트가 아니라 답변들 일급컬렉션을 파라미터로 갖도록 수정
질문을 삭제하면, 연관된 답변의 삭제처리 및 삭제이력 생성되도록 수정

Commit:
feat: 사용자 클래스 리팩토링
삭제이력 리스트가 아니라 삭제이력들들 일급컬렉션을 파라미터로 갖도록 수정
삭제이력의 삭제자랑 현재 추가하려는 사용자가 일치해야만 추가할 수 있도록 검증로직 추가

Commit:
refactor: 답변 테스트코드 리팩토링.
1. 인자를 private으로 수정
2. 답변의 삭제를 답변 도메인 안으로 가져옴
3. 답변을 삭제하면 삭제이력이 반환되도록 구현

Commit:
test: 답변 컬렉션 테스트코드 개발
1. 전체 삭제와 관련된 테스트로직
2. 답변 추가와 관련된 테스트로직

Commit:
refactor: 질문 테스트코드 리팩토링
1. 질문에 답변 추가하는 로직 테스트
2. 질문의 삭제와 관련된 로직 테스트

Commit:
refactor: 사용자 테스트코드 리팩토링
1. 프라이빗 파라미터로 변경
2. 삭제이력 관련 로직 추가

Commit:
refactor: 레포지토리 테스트코드 추가 - 마이너 변경
엔티티 클래스의 수정으로 인한 레포지토리 테스트 수정
레포지토리 함수 변경된 부분 없으므로 큰 변경 x

Commit:
refactor: QnA 서비스 코드 리팩토링
1. 불필요한 AnswerRepository 의존성 삭제
2. 삭제로직 각 도메인 로직 안으로 들여보냄

Commit:
docs: 리드미 기능 구현 완료된 내용 체크

heowc commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -89,14 +87,51 @@ public void setDeleted(boolean deleted) {
                 "id=" + id +
-                ", writer=" + writer.toString() +
-                ", question=" + question.toString() +
+                ", writer=" + writer +
```
앗, 저는 https://lng1982.tistory.com/300 이러한 이유 떄문에 언급을 드린거였는데 잘 못 얘기했네요 ㅎㅎ;;

badaelephant replied:
안녕하세요 원철님. JPA에서 직렬화 순환참조 문제를 해결하는 방법중 좋다고 판단되는 방법이 아래 2가지 있는데요, 
@JsonManagedReference, @JsonBackReference
@JsonIdentityInfo
현재 프로젝트 환경에서는 해당 어노테이션이 없는데 혹시 어떤 패키지를 추가해야 하는 것일까요?

heowc replied:
의존성 추가를 지양하고 있기 때문에 id만 출력하는 것으로 마무리해볼 수 있지 않을까 싶습니다 🤔

heowc commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,43 @@
+                        .collect(Collectors.toList()));
+    }
+
+    public void update(Answer answer) {
```
update 보다는 add는 어떨까요 ? 🤔

heowc commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,43 @@
+    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL)
+    private List<Answer> answers;
+
+    public Answers(List<Answer> answers) {
```
생성시 또는 내부 콜렉션 객체를 반환시 방어적복사를 하는건 어떨까요? 🤔 
- https://alexander96.tistory.com/16

badaelephant replied:
안녕하세요 원철님. 이부분 수정했는데, 혹시 맞게 수정되었다고 생각되시는지 한번 확인 부탁드립니다.

heowc commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,47 @@
+public class DeleteHistories {
+
+    @OneToMany(mappedBy = "deletedBy", cascade = CascadeType.ALL)
+    private List<DeleteHistory> deleteHistoryList;
```
특정 구현체(ex. List)가 포함된 변수명은 지양하는 것이 좋다고 생각합니다. 
복수형의 의미로 `deleteHistories`는 어떨까요 ? 🤔

heowc commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,47 @@
+
+    }
+
+    public static DeleteHistories of(DeleteHistory questionDeleteHistory, DeleteHistories answerDeleteHistories) {
```
`DeleteHistories` 생성하는 곳이 두군데로 나뉘어 지는것이 다소 어색하다고 느꼈는데요.
보통 정적 팩토리 메소드에 대한 메소드명을 작성할 때, `of`, `from`을 사용하기 때문이라고 생각합니다.

이를 어떻게 개선해볼 수 있을까요? 🤔

heowc commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,47 @@
+        this.deleteHistoryList.add(deleteHistory);
+    }
+
+    public int getSize() {
```
`getSize()`를 사용하는 곳을 보니 테스트에서만 사용하는 것을 볼 수 있었습니다.
그리고 관련테스트가 꼭 필요한지 의문이 들었는데요?

어떻게 생각하시나요 ? 🤔

heowc commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -42,26 +30,7 @@ public Question findQuestionById(Long id) {
-                    new DeleteHistory(ContentType.ANSWER, answer.getId(), question.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = question.delete(loginUser);
```
깔끔한 리팩토링 👍

heowc reviewed (changes requested):
바쁘신 와중에도 빠르게 3단계를 진행해주니다니.. 고생이 많으시네요 👍 

해당 단계를 잘 진행해주셨지만, 마지막 단계인 만큼 소소한 피드백을 남겨보았습니다.
궁금하신 점이나 이야기 나누고 싶은 부분이 있다면 깃헙 코멘트, 슬랙 DM 편하게 남겨주세요! 🙏 

--- 

> 작성하면서 조금 궁금했던 것이 한가지 있는데, 바로 DeleteHistories 등의 getter함수에 대한 부분이었습니다.

저같은 경우는 그래서 JpaRepository까지 DeleteHistories를 전달해주고 거기서  `deleteHistories.getValue()` 리스트를 반환했던 기억이 납니다. 대신 `getValue()`에 접근제한자를 default로 해서요.
저도 많은 걸 찾아보진 않았지만, 말씀하신 것처럼 로 인해 Entity가 아니라서 어려움이 있는 것 같아요. 대신 저와 비슷하게 최대한 내부에 있는 컬랙션을 최대한 덜 접근할 수 있는 방법을 모색해보면 좋지 않을까 싶습니다 👍 
답변이 되었을지 모르겠네요... 🙏

Commit:
refactor: 피드백 반영
1. DeleteHistory.of 정적 팩토리 메소드 명을 목적에 맞게 명확히 함
2. toString 무한참조를 막기위해 question을 직접 받지 않고 question_id를 받음
3. List에 Item을 추가할때 update대신 add라는 메소드명으로 작성
4. 컬렉션에 방어적 복사 추가

heowc commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -115,20 +115,22 @@ public boolean equals(Object o) {
                 "id=" + id +
-                ", writer=" + writer +
-                ", question=" + question +
+                ", writer=" + writer.toString() +
```
이것도 `.getId()` 하는게 좋지 않을까요 !? 🤔

heowc commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -11,37 +11,27 @@
+        this.deleteHistories.addAll(deleteHistories.getList());
     }
 
     public boolean contains(DeleteHistory deleteHistory) {
```
지금보니 해당 메소드도 테스트에만 사용되고 있네요.
개인적인 생각은 객체가 정상적으로 만들어졌다는 것을 확인하는 것만으로도 충분한 테스트가 되지 않을까싶습니다 🤔 

테스트 범위가 깊어질 수록 오히려 테스트하고자 하는 목적(?) 방향성(?)을 잃어버릴 수 도 있다고 생각하는데요.
어떻게 생각하시나요!? 🤔

heowc commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -35,11 +35,11 @@ private DeleteHistory(ContentType contentType, Long contentId, User deletedBy, L
     }
 
-    public static DeleteHistory of(Question question) {
+    public static DeleteHistory ofQuestion(Question question) {
```
보다 명시적이네요 👍

heowc commented on src/main/java/qna/domain/Question.java:
```diff
@@ -92,10 +91,12 @@ public void setDeleted(boolean deleted) {
+        DeleteHistory questionDeleteHistory = DeleteHistory.ofQuestion(this);
         this.deleted = true;
-        return DeleteHistories.of(questionDeleteHistory, answerDeleteHistories);
+        DeleteHistories deleteHistories = new DeleteHistories();
```
`DeleteHistories`를 만들고 각각의 히스토리를 추가하는 방식으로 진행하셨네요 👍

heowc commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -40,4 +37,6 @@ public boolean contains(Answer answer) {
     private List<Answer> findDeletableAnswers() {
         return answers.stream().filter(ans -> !ans.isDeleted()).collect(Collectors.toList());
     }
+
+
```
```suggestion
```

불필요한 라인은 제거해주세요 👍

heowc reviewed (approved):
마지막까지 고민 많이 하시고 피드백도 반영해주셔서 감사합니다.

소소한 코멘트를 남겼으니 확인해주세요.
다만 merge하기엔 충분하여 이만 미션은 마치도록 할게요 👍 
고생 많으셨어요 🙇 💯 

궁금하신 점이나 의논할 것이 있다면 깃헙 코멘트나 슬랙 DM 으로 연락주세요 🙏
